### PR TITLE
Optimise redbank liquidations

### PR DESCRIPTION
**Pull request by markonmars**
TL:DR 
- Prevent unnecessary swaps when incorrectly attempting to liquidate a single redbank position.
- optimise fetching of data to reduce time from identification to liquidation
- add new test for single liquidation, and follow same patterns as rover

**Context**:

When liquidating on redbank, we use a proxy contract which takes liquidation messages and feeds them to the redbank on our behalf, allowing us to liquidate multiple accounts in a single transaction safely. 

For instance, consider the following : 

- There are 10 positions eligible for liquidation
- We send a transaction with 10 liquidation messages to the redbank
- One of those positions is now healthy (other liquidator, price change etc). 

This will cause the entire tx to fail, even if 9 of the 10 were valid. To solve this we use the filterer contract.

However, this presents an issue, as transactions will succeed when even **ZERO** positions are healthy.

Because we dispatch the swap from our stable to the debt asset **before** the liquidation in the same transaction, if there is no liquidation possible we do not want to succeed with the transaction, as we will have to pay for the swap, and then swap back afterwards despite winning no collateral. This is costly, especially for large positions. 

By using redbank where possible, we will fail if the position is not liquidatable.

Ultimately this is not a perfect solution - it is better for the filterer to fail the tx if there are no successful liquidations, or alternatively (prefered imo) we do not use the filterer and instead submit multiple independent atomic tx's.

### CODE REVIEW

**Review by elfedy — APPROVED** [review]
Looks good a couple of thoughts here:

I remember back in the day we were trying to pack liquidation + swaps in the liquidations helper contract: https://github.com/mars-protocol/mars-liquidator/pull/2 that would sort of solve the issue but it's putting complexity on the SC so there's a tradeoff to be made.


I think a better way to handle this is to just not do swaps at all during liquidations. You would have X% of your funds in the "neutral" denom and Y% spread among the other denoms.

There could be several rules to determine this for example:

* % Of debt in each asset gets % of allocation (with some allowed variance)
* Have enough liquidity of each denom so that  we could liquidate the 20 most risky positions assuming it's health factor <= 1 (so half of the current debt of the position or something like that)

This way there are two benefits:

1. You don't have this swapping logic on the liquidation anymore.
2. Liquidator liquidates so it's not awaiting for the swaps to come back to continue the liquidation loop
3. You save lots of fees.

**Comment by markonmars** [comment]
Yea, I pushed to have funds and all swap logic in the filterer contract initially (then the bot just calls `liquidate(addresses : string[])`), but there were a few reasons to not do that (primarily add complexity on chain + smart contract dev resources scarce) so in the end we just did the filterer.

> I think a better way to handle this is to just not do swaps at all during liquidations

I have also thought of something similar. Initial implementation was more like this but the key problem is balancing the assets and how you balance them- You would still need some sort of logic that rebalances the assets based on some rules like you describe above, so that you can meet liquidation requirements. I imagine you would require mostly stable or mostly volatile assets depending on which way the market is moving.

The liquidations for rover are much nicer as we don't need bot capital, and we can make use of the `AccountBalance` feature to swap / withdraw /repay everything without needing to specify amounts. It would be nice if we could add a `LiquidateRedbank` action to rover but I don't understand what the trade offs are here with complexity / difficulty etc
